### DISCUSSION TITLE
NeoPixel write freezes ESP32-S2, S3, C3

### DISCUSSION THREAD

**Discussion by rdagger** [micropython/micropython#15346] [ESP32]
The ESP32-S2 will eventually lock up if you don't put at least a 15ms sleep between `neopixel.fill` and `neopixel.write`. This issue has also been reported on the S3 and C3. There's an [open issue](https://github.com/micropython/micropython/issues/10554) about it that was posted a year and a half ago.

In my project, timing is critical, and I can't afford a 15ms sleep in my main loop. I tried using a state machine to space out the fill and write methods at 15ms intervals, but it didn't work. It seems only a sleep right before write prevents the freezing.

I'm running ESP32_GENERIC_S2-20240602-v1.23.0. on a Qt Py ESP32-S2.

Does anyone have a workaround?

**Comment by ctimmer** [comment]
No expertise here but...
Because you are not using sleep maybe you need more than 15ms after the fill.

Curt

**Reply by rdagger** [reply]
> No expertise here but... Because you are not using sleep maybe you need more than 15ms after the fill.
> 
> Curt

I increased the interval to 2000ms, and it still did not work.  It's not the gap between the fill and write.  Apparently, the write method needs spacing between other operations too.

Here's my state machine code I'm using to troubleshoot:
```
from machine import Pin  # type: ignore
from neopixel import NeoPixel  # type: ignore
from time import sleep_ms, ticks_ms, ticks_diff, ticks_add
np = NeoPixel(Pin(39), 1)  # NeoPixel LED
np_power = Pin(38, Pin.OUT)  # NeoPixel Power Pin
np_power.on()  # Enable NeoPixel power


class NeoPixelFader:
    COLORS = (
        (128, 0, 255),   # Indigo
        (255, 0, 255),   # Magenta
        (255, 0, 128),   # Fuchsia
        (255, 0, 0),     # Red
        (255, 128, 0),   # Orange
        (255, 255, 0),   # Yellow
        (128, 255, 0),   # Chartreuse
        (0, 255, 0),     # Green
        (0, 255, 128),   # Aqua
        (0, 255, 255),   # Cyan
        (0, 128, 255),   # Sky Blue
        (0, 0, 255)    # Blue
    )
    DELAY = 500  # Ms delay between fill and write

    def __init__(self, neopixel):
        self.np = neopixel
        self.state = 0
        self.index = 0
        self.next_time = ticks_ms()

    def update(self):
        current_time = ticks_ms()
        if self.state == 0:
            self.np.fill(self.COLORS[self.index])
            self.next_time = ticks_add(current_time, self.DELAY)
            self.state = 1
        elif self.state == 1:
            if ticks_diff(current_time, self.next_time) >= 0:
                self.state = 0
                self.index = (self.index + 1) % len(self.COLORS)
                # sleep_ms(15)  # Required or board freezes
                self.np.write()


fader = NeoPixelFader(np)
try:
    print('Running...')
    while True:
        fader.update()
except KeyboardInterrupt:
    print("Ctrl-C pressed.")
finally:
    np.fill((0, 0, 0))   # Turn off NeoPixel
    sleep_ms(15)
    np.write()  # Update NeoPixel
    print('Done.')
```

It crashes after about 1 loop.

**Comment by rdagger** [comment]
Here's a very simple program to reproduce the crash:

```
from machine import Pin  # type: ignore
from neopixel import NeoPixel  # type: ignore
from time import sleep_ms
np = NeoPixel(Pin(39), 1)  # NeoPixel LED
np_power = Pin(38, Pin.OUT)  # NeoPixel Power Pin
np_power.on()  # Enable NeoPixel power

COLORS = (
    (128, 0, 255),   # Indigo
    (255, 0, 255),   # Magenta
    (255, 0, 128),   # Fuchsia
    (255, 0, 0),     # Red
    (255, 128, 0),   # Orange
    (255, 255, 0),   # Yellow
    (128, 255, 0),   # Chartreuse
    (0, 255, 0),     # Green
    (0, 255, 128),   # Aqua
    (0, 255, 255),   # Cyan
    (0, 128, 255),   # Sky Blue
    (0, 0, 255)    # Blue
)
DELAY = 20  # 20 ms works, 10 ms does not work


try:
    print('Running...')
    while True:
        for color in COLORS:
            np[0] = color
            sleep_ms(DELAY)
            np.write()
            sleep_ms(500)
except KeyboardInterrupt:
    print("Ctrl-C pressed.")
finally:
    np[0] = (0, 0, 0)
    sleep_ms(DELAY)
    np.write()
    print('Done.')
```

**Reply by ricksorensen** [reply]
I am running on a UM TinyS3 (data:GPIO18, pwr:GPIO17), onboard 2812 )  with 5ms sleep .,.. no problems.
     `(name='micropython', version=(1, 23, 0, 'preview'), _machine='TinyS3 with ESP32-S3-FN8', _mpy=11014)`
And on a XIAO ESP32C3 (data:GPIO2, constant power, external strip 2811) with no sleep ... no problems
     `(name='micropython', version=(1, 23, 0, 'preview'), _machine='ESP32C3 XIAO with ESP32C3', _mpy=518)`

Running with the 1.23 release version of GENERIC_C3 with no sleep... again no problems
` (name='micropython', version=(1, 23, 0, ''), _machine='ESP32C3 module with ESP32C3', _mpy=774)`

**Comment by rdagger** [comment]
I guess the generic v1.23.0 must not work with the QT-PY ESP's.  I tried the same code in CircuitPython and it works. 

Do you think one of the other S2 distro would work better such as Feather S2, Feather S2 Neo, Tiny S2, S2 Mini or S2 Pico?

Is there a custom build setting I need?

**Comment by steveh127** [comment]
I see no issues with neopixel without any delay using XIAO S3 with generic S3 build on any build I've used upto 1.24 preview. Nor with any Unexpected maker ESP32 S3 board using targeted builds. Perhaps the workaround is to use a different board?

**Comment by rdagger** [comment]
I made a mistake.  The simple test to reproduce the problem above turns out to work most of the time without a delay.  But the slightly more complicated program below always fails after a few loops unless a 20ms sleep is added.
 
```
from machine import Pin  # type: ignore
from neopixel import NeoPixel  # type: ignore
from time import sleep_ms
np = NeoPixel(Pin(39), 1)  # NeoPixel LED
np_power = Pin(38, Pin.OUT)  # NeoPixel Power Pin
np_power.on()  # Enable NeoPixel power

def hsv_to_rgb(h, s, v):
    """
    Convert HSV to RGB (based on colorsys.py).

        Args:
            h (float): Hue 0 to 1.
            s (float): Saturation 0 to 1.
            v (float): Value 0 to 1 (Brightness).
    """
    if s == 0.0:
        return v, v, v
    i = int(h * 6.0)
    f = (h * 6.0) - i
    p = v * (1.0 - s)
    q = v * (1.0 - s * f)
    t = v * (1.0 - s * (1.0 - f))
    i = i % 6
    v = int(v * 255)
    t = int(t * 255)
    p = int(p * 255)
    q = int(q * 255)
    if i == 0:
        return v, t, p
    if i == 1:
        return q, v, p
    if i == 2:
        return p, v, t
    if i == 3:
        return p, q, v
    if i == 4:
        return t, p, v
    if i == 5:
        return v, p, q


DELAY = 20
spectrum = tuple(range(512))
try:
    print('Running...')
    while True:
        for c in spectrum:
            hue = c / 512.0
            rgb = hsv_to_rgb(hue, 1, .15)
            print(rgb)
            np[0] = rgb
            #sleep_ms(DELAY)
            np.write()
except KeyboardInterrupt:
    print("Ctrl-C pressed.")
finally:
    np[0] = (0, 0, 0)  # Turn off NeoPixel
    sleep_ms(DELAY)
    np.write()  # Update NeoPixel
    print('Done.')
```

**Reply by ricksorensen** [reply]
My `tinyS3` ran with no problems.
My `XIAO ESP32C3` hung after a few iterations, `(23,38,0)` was the last rgb printed

Testing with ESP32C3:
Going back to the 'simple' example, and removing the 500ms sleep after the `np.write() `  (no sleep before write) the computer hung.
Setting the sleep after `np.write` to 5 the system hangs
Setting the sleep after `np.write` to 10 and it keeps running.

It had not failed with a `XIAO RP2040`

The `neopixel` module itself is pure python, but it imports `machine.bitstream` --- maybe this needs looking at? 

I do not yet have great ESP32 debugging skills- maybe an expert can chime in?

**Comment by rdagger** [comment]
I tested a Qt-Py ESP32-S3 and it works without any delays for both tests. I installed ESP32_GENERIC_S3-20240602-v1.23.0.

It seems the bug is specific to the S2 and C3.

**Reply by ricksorensen** [reply]
I just ran an experiment with my XIAO C3 building with different ESP compilers:
- V5.04, which I think is the current release version, the program runs briefly then hangs.  I do not know what 'hang' means exactly but REPL is not responsive to any keys including `Ctrl-C,Ctrl-D,Ctrl-X`
- v5.1 program runs without issue
- v5.2.1 program runs without issue

Maybe something in the 5.04 SDK?
